### PR TITLE
Support recursive CTEs

### PR DESCRIPTION
**Pull request by kasiafi**
Adds support for recursive common table expression (`WITH RECURSIVE`).
Single-element recursive cycles are allowed.

Issue: #1122

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -2515,42 +2534,274 @@ private void analyzeColumnMask(String currentIdentity, Table table, QualifiedObj
...
-                }
+                                List<Type> anchorFieldTypes = anchorType.getVisibleFields().stream()
+                                        .map(Field::getType)
+                                        .collect(toImmutableList());
+                                List<Type> stepFieldTypes = stepType.getVisibleFields().stream()
+                                        .map(Field::getType)
+                                        .collect(toImmutableList());
+                                if (!anchorFieldTypes.equals(stepFieldTypes)) {
+                                    analysis.addRelationCoercion(step, anchorFieldTypes.toArray(Type[]::new));
+                                }
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java:2641]: Move this below the block that checks for type compatibility. It doesn't change the outcome, but it's "logically" more correct.

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 2]
```diff
@@ -2515,42 +2534,274 @@ private void analyzeColumnMask(String currentIdentity, Table table, QualifiedObj
...
+
+                                for (int i = 0; i < anchorFieldTypes.size(); i++) {
+                                    if (!typeCoercion.canCoerce(stepFieldTypes.get(i), anchorFieldTypes.get(i))) {
+                                        throw semanticException(
+                                                TYPE_MISMATCH,
+                                                step,
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java:2647]: Improvement for the future: use the corresponding SELECT expression as the "node" in the exception so that error highlighting points at the right place. It won't be trivial to do, so let's defer it (add a TODO)

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 3]
```diff
@@ -2515,42 +2534,274 @@ private void analyzeColumnMask(String currentIdentity, Table table, QualifiedObj
...
-                // check if all or none of the columns are explicitly alias
-                if (withQuery.getColumnNames().isPresent()) {
-                    List<Identifier> columnNames = withQuery.getColumnNames().get();
-                    RelationType queryDescriptor = analysis.getOutputDescriptor(query);
-                    if (columnNames.size() != queryDescriptor.getVisibleFieldCount()) {
-                        throw semanticException(MISMATCHED_COLUMN_ALIASES, withQuery, "WITH column alias list has %s entries but WITH query(%s) has %s columns", columnNames.size(), name, queryDescriptor.getVisibleFieldCount());
+                                analysis.setScope(withQuery.getQuery(), aliasedAnchorScope);
+                                analysis.registerExpandableQuery(withQuery.getQuery(), recursiveReference);
+                                withScopeBuilder.withNamedQuery(name, withQuery);
+                                continue;
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java:2658]: The many levels of nesting and this `continue` here make it hard to follow the flow of the code. One possible way of reorganizing this is to move this whole section to a new method and have it return "true" if it was able to handle the recursive query. Then, the code below would do:

```
if (!handled) {
   // 2. 
   ...
   // process plain query
   ...
}
```

Let me play with the code a bit to see if I can come up with something better.

Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java:2658]: Also, instead of splitting the method into two sections based on `with.isRecursive()`, it seems we should be able to structure in a similar manner to this and avoid a bit of duplication:

```java
With with = node.getWith().get();
Scope.Builder withScopeBuilder = scopeBuilder(scope);

for (WithQuery withQuery : with.getQueries()) {
    String name = withQuery.getName().getValue().toLowerCase(ENGLISH);
    if (withScopeBuilder.containsNamedQuery(name)) {
        throw semanticException(DUPLICATE_NAMED_QUERY, withQuery, "WITH query name '%s' specified more than once", name);
    }

    boolean isRecursive = false;
    if (with.isRecursive()) {
        isRecursive = tryProcessRecursiveQuery(...);
    }

    if (!isRecursive) {
        // check if all or none of the columns are explicitly alias
        Query query = withQuery.getQuery();
        if (withQuery.getColumnNames().isPresent()) {
            validateColumnAliases(withQuery.getColumnNames().get(), analysis.getOutputDescriptor(query).getVisibleFieldCount());
        }

        process(query, withScopeBuilder.build());
        withScopeBuilder.withNamedQuery(name, withQuery);
    }
}

Scope withScope = withScopeBuilder.build();
analysis.setScope(with, withScope);
return withScope;
```

Possibly, even this for the inner condition, but I'm not a big fan of the side-effect in the `if` condition:

```java
...
if (!with.isRecursive() || !tryProcessRecursiveQuery(...)) {
   ...
}
...
```

File: presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java [hunk 4]
```diff
@@ -161,6 +185,172 @@ public RelationPlan plan(Query query)
...
+                ImmutableList.of(),
+                AggregationNode.Step.SINGLE,
+                Optional.empty(),
+                Optional.empty());
+
+        return new RelationPlan(
+                union.isDistinct() ? distinctAggregation : unionNode,
+                anchorPlan.getScope(),
+                unionOutputSymbols,
+                outerContext);
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java:316]:
```suggestion
        PlanNode result = new UnionNode(idAllocator.getNextId(), recursionStepsToUnion, unionSymbolMapping.build(), unionOutputSymbols);

        if (union.isDistinct()) {
            result = new AggregationNode(
                    idAllocator.getNextId(),
                    result,
                    ImmutableMap.of(),
                    singleGroupingSet(result.getOutputSymbols()),
                    ImmutableList.of(),
                    AggregationNode.Step.SINGLE,
                    Optional.empty(),
                    Optional.empty());
        }

        return new RelationPlan(result, anchorPlan.getScope(), unionOutputSymbols, outerContext);
```

File: presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java [hunk 5]
```diff
@@ -114,6 +114,7 @@
     private final Optional<TranslationMap> outerContext;
     private final Session session;
     private final SubqueryPlanner subqueryPlanner;
+    private final Map<NodeRef<Node>, RelationPlan> replacementMap;
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java:117]: `replacementMap` is too vague of a name. Maybe rename to something more precise, such as "recursiveSubqueries".

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java (hunk 2 above)
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java:2647]: The number of the first mismatching column is included in the error message. It is not always possible to highlight the exact select expression, because it might be e.g. a "row all fields" expression. I'm not sure what I should write in the TODO.

File: presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java [hunk 6]
```diff
@@ -161,6 +185,172 @@ public RelationPlan plan(Query query)
...
+        // plan anchor relation
+        Relation anchorNode = union.getRelations().get(0);
+        RelationPlan anchorPlan = new RelationPlanner(analysis, symbolAllocator, idAllocator, lambdaDeclarationToSymbolMap, metadata, outerContext, session, replacementMap)
+                .process(anchorNode, null);
+
+        // prune anchor plan outputs to contain only the symbols exposed in the scope
+        List<Symbol> anchorOutputs = anchorPlan.getRoot().getOutputSymbols().stream()
+                .filter(ImmutableSet.copyOf(anchorPlan.getFieldMappings())::contains)
+                .collect(toImmutableList());
+        PlanNode anchorPlanRoot = new ProjectNode(idAllocator.getNextId(), anchorPlan.getRoot(), Assignments.identity(anchorOutputs));
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java:204]: This is necessary for successful replacement. At some point we must plug recursion step plan in the place where anchor plan was.
As both anchor plan and step plan might have additional output symbols, they need to be pruned so that the replacement fits and the output symbols match their "continuations".

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/AbstractSymbolMapper.java [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.prestosql.sql.planner.plan.AggregationNode.groupingSets;
+import static java.util.Objects.requireNonNull;
+
+public abstract class AbstractSymbolMapper
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/AbstractSymbolMapper.java:53]: Using inheritance to share code is generally not a good approach. It introduces issues of poor encapsulation, unclear thread-safety semantics (for code that can be accessed concurrently), etc. It's better to do it via delegation. In this case, you could make SymbolMapper take an object during construction that is responsible for doing the primordial mapping. You'd customize that for each of the two use cases we need here.

File: presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java [hunk 8]
```diff
@@ -198,14 +198,10 @@ public RelationPlan planExpand(Query query)
                 .process(anchorNode, null);
 
         // prune anchor plan outputs to contain only the symbols exposed in the scope
-        List<Symbol> anchorOutputs = anchorPlan.getRoot().getOutputSymbols().stream()
-                .filter(ImmutableSet.copyOf(anchorPlan.getFieldMappings())::contains)
-                .collect(toImmutableList());
-        PlanNode anchorPlanRoot = new ProjectNode(idAllocator.getNextId(), anchorPlan.getRoot(), Assignments.identity(anchorOutputs));
-
-        anchorPlan = new RelationPlan(anchorPlanRoot, analysis.getScope(query), anchorOutputs, outerContext);
+        NodeAndMappings prunedAnchorPlan = pruneInvisibleFields(anchorPlan, idAllocator);
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java:201]: It is necessary to prune anchor outputs. If we leave those extra symbols in the anchor plan, they will "leak" to the recursion step plan. Later, when anchor plan is replaced with step plan, they will become orphaned.

File: presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java [hunk 9]
```diff
@@ -232,26 +228,28 @@ public RelationPlan planExpand(Query query)
...
         PlanNode replacementSpot = anchorPlan.getRoot();
         PlanNode recursionStep = coercedRecursionStep.getNode();
+        List<Symbol> mappings = coercedRecursionStep.getFields();
 
         // unroll recursion
         int maxRecursionDepth = getMaxRecursionDepth(session);
         for (int i = 0; i < maxRecursionDepth; i++) {
-            recursionSteps.add(copy(recursionStep));
-            PlanNode replacement = copy(recursionStep);
+            recursionSteps.add(alignOutputs(copy(recursionStep, mappings)));
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java]: I changed it so that mappings are passed along with the PlanNode and whenever we need to bind symbols properly (that is, when replacing and when building UnionNode), it is done via mapping.

**Comment by kasiafi** [comment]
Updated and squashed.
